p11-kit/proxy.c

```diff
@@ -1785,11 +2285,153 @@ C_GetFunctionList (CK_FUNCTION_LIST_PTR_PTR list)
 	}
 
 	if (rv == CKR_OK)
-		*list = module;
+		*list = (void *)module;
 
 	if (loaded)
 		p11_kit_modules_release (loaded);
 
+	return rv;
+}
+
+static const CK_VERSION version_two = {CRYPTOKI_LEGACY_VERSION_MAJOR, CRYPTOKI_LEGACY_VERSION_MINOR};
+
+/* We are not going to support any special interfaces */
+#define NUM_INTERFACES 2
+#define DEFAULT_INTERFACE 0
+CK_INTERFACE interfaces[NUM_INTERFACES] = {
+        {"PKCS 11", NULL, 0}, /* 3.0 */
+        {"PKCS 11", NULL, 0}  /* 2.4 */
+};
+
+static int
+internal_init_interfaces (void)
+{
+	int rv = CKR_ARGUMENTS_BAD;
+	void *res = NULL;
+
+	/* Version 3.0 is default */
+	rv = get_function_list_inlock (&res, NULL);
+	if (rv != CKR_OK)
+		return rv;
+	interfaces[0].pFunctionList = res;
+
+	rv = get_function_list_inlock (&res, &version_two);
+	if (rv != CKR_OK)
+		return rv;
+	interfaces[1].pFunctionList = res;
+
+	return CKR_OK;
+}
+
+#ifdef OS_WIN32
+__declspec(dllexport)
+#endif
+CK_RV
+C_GetFunctionList (CK_FUNCTION_LIST_PTR_PTR list)
+{
+	CK_RV rv = CKR_OK;
+	void *res = NULL;
+
+	p11_library_init_once ();
+	p11_lock ();
+
+	rv = get_function_list_inlock (&res, &version_two);
+	if (rv == CKR_OK)
+		*list = res;
+
+	p11_unlock ();
+
+	return rv;
+}
+
+#ifdef OS_WIN32
+__declspec(dllexport)
+#endif
+CK_RV
+C_GetInterfaceList (CK_INTERFACE_PTR pInterfacesList, CK_ULONG_PTR pulCount)
+{
+	CK_RV rv = CKR_OK;
+
+	if (pulCount == NULL_PTR)
+		return CKR_ARGUMENTS_BAD;
+
+	if (pInterfacesList == NULL_PTR) {
+		*pulCount = NUM_INTERFACES;
+		return CKR_OK;
+	}
+
+	if (*pulCount < NUM_INTERFACES) {
+		*pulCount = NUM_INTERFACES;
+		return CKR_BUFFER_TOO_SMALL;
+	}
+
+	p11_library_init_once ();
+	p11_lock ();
+
+	rv = internal_init_interfaces ();
+	if (rv == CKR_OK) {
+		memcpy (pInterfacesList, interfaces, NUM_INTERFACES * sizeof(CK_INTERFACE));
```

Comment:
Can't `internal_init_interfaces` be extended to do this `memcpy` internally?

Reply:
I wouldnt do that. Right now the init_interfaces function does exactly what it is supposed to: initializes internal interfaces array. I think it makes more sense this way.

Reply:
If we do this way (i.e., copy the result), we don't need to define `interfaces` as global; it can be a local variable in this function.

Reply:
I am keeping it as is. I cant make it into local variable without reworking a lot more then just `init_interfaces`, for example this part inside `C_GetInterface` expects it to be global as well:
`*ppInterface = &interfaces[i];`